### PR TITLE
feat: Add bare Kotlin type overloads for expression functions

### LINKED ISSUE #193: Add bare Kotlin type overloads for expression functions

### PR DESCRIPTION
**Pull request by sargunv**
closes #193

### CODE REVIEW

**Comment by sargunv** [comment]
* **#196** <a href="https://app.graphite.dev/github/pr/sargunv/maplibre-compose/196?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/sargunv/maplibre-compose/196?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/math.kt [hunk 1]
```diff
@@ -9,13 +9,13 @@ import dev.sargunv.maplibrecompose.expressions.value.NumberValue
...
-public fun ln2(): Expression<FloatValue> = FunctionCall.of("ln2").cast()
+public val LN_2: Expression<FloatValue> = FunctionCall.of("ln2").cast()
 
 /** Returns the mathematical constant π */
-public fun pi(): Expression<FloatValue> = FunctionCall.of("pi").cast()
+public val PI: Expression<FloatValue> = FunctionCall.of("pi").cast()
 
 /** Returns the mathematical constant e */
-public fun e(): Expression<FloatValue> = FunctionCall.of("e").cast()
+public val E: Expression<FloatValue> = FunctionCall.of("e").cast()
```
Comment (sargunv) [inline lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/math.kt:18]: tweaking these to match conventions in `kotlin.math`

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/variables.kt [hunk 2]
```diff
@@ -13,15 +13,12 @@ import kotlin.jvm.JvmInline
  *   x.use() + const(3)
  * }
  * ```
- *
- * Variable names starting with `__` are reserved for internal use by the library.
```
Comment (sargunv) [inline lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/variables.kt:17]: this is no longer needed with the new AST

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/math.kt
Comment (westnordost) [inline lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/math.kt:1]: all overloads with bare types are missing for the math functions

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/ramps.kt
Comment (westnordost) [inline lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/ramps.kt:1]: The stops for the interpolate expressions and step expressions don't have bare Kotlin overloads.

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/feature.kt [hunk 3]
```diff
@@ -43,10 +52,27 @@ public object Feature {
    * or any primitive data type. Note that [state] can only be used with layer properties that
    * support data-driven styling.
    */
-  // TODO: document which layer properties support feature state expressions on which platforms
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/feature.kt:46]: I replaced the TODO with an issue instead: #195

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/decision.kt
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/decision.kt:1]: It's just stylistic preference. I prefer to think of the expressions DSL as regular functions, and short-lived objects like `Case` and `Condition` are implementation details to get the typing and syntax right.

Additionally, Dokka shows documentation for functions on a different page from types, and this keeps documentation for the construct and also the branches on the same page.

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/ramps.kt
Comment (sargunv) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/ramps.kt:1]: With current typing (the `Pair`), either all stops will have to be Kotlin values, or all stops Expressions. Thoughts on doing something like `case` and `condition` (maybe `stop`?) so we can use different overloads on each stop?

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/math.kt
Comment (sargunv) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/math.kt:1]: For math operators, it felt odd to me to have different types on two sides of the operator. Maybe it's okay though

Comment (sargunv) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/math.kt:1]: also for anything related to numbers with units, we'd write a lot of overloads for bare types (no unit, Dp, TextUnit, Duration, maybe Meters in the future). Seems not worth it?

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/collections.kt [hunk 4]
```diff
@@ -16,6 +16,11 @@ public operator fun <T : ExpressionValue> Expression<ListValue<T>>.get(
...
 ): Expression<T> = FunctionCall.of("at", index, this).cast()
 
+/** Returns the item at [index]. */
+@JvmName("getAt")
+public operator fun <T : ExpressionValue> Expression<ListValue<T>>.get(index: Int): Expression<T> =
+  get(const(index))
+
 /** Returns whether this list contains the [item]. */
 @JvmName("containsList")
 public fun <T : ExpressionValue> Expression<ListValue<T>>.contains(
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/collections.kt:26]: what would be the most useful overloads here? I could see overloads on the left for List<String>, List<Boolean>, List<Number> and then on the right for String, Boolean, Number, so six in total?

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/decision.kt
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/decision.kt:1]: and similar to math, overloads for generic stops means writing a whole lot of overloads for many different types

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/collections.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (westnordost) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/collections.kt:26]: most expressions that use `contains` check if the value of some feature property is in a set of values, so e.g. `listOf("primary", "secondary", "tertiary").contains(feature["highway"])`

Can't think of a use case for having the right side as a literal.

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/decision.kt
[the remark this answers is not in the document]
Comment (westnordost) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/decision.kt:1]: true... , but using literals for these constructs is pretty much the default case for these constructs.

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/math.kt
Comment (westnordost) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/math.kt:1]: Well, not sure, IIRC the Android MapLIbre SDK does that. (For Floats only, of course)

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/ramps.kt
Comment (westnordost) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/ramps.kt:1]: I think it is rather the default / usual case that no expressions are used, so the overload that doesn't allow any expressions for the stops would I think be the most used one.
Having data classes for `step` and `interpolate` i.e. `stop(...)` sounds fine to me, too. It's only marginally less convenient/more verbose than the `to` (and anyway, that is just a normal infix function, you could name the function that creates a `Stop` also like that.)

Comment (sargunv) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/ramps.kt:1]: I'd be wary of shadowing `to` so would want a different infix name, or just `stop()` like the others. 

It's true that most stops/case/condition would be `const`, but also interpolate/switch is already a pretty big, verbose, construct, and so adding one const per stop doesn't hurt readability much (at least to my eye).

Compared to `get` and `has` which were small enough that the const dwarfed the get/has and made it confusing to read.

File: lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/collections.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose-expressions/src/commonMain/kotlin/dev/sargunv/maplibrecompose/expressions/dsl/collections.kt:26]: Adding an overload for List<String> sounds good to me if it doesn't shadow the regular Kotlin `contains`. 

If it does shadow, well I think we already have `const(listOf())` so it's not too bad. Should there be a `constList(varargs)`?

**Comment by westnordost** [comment]
it wouldn't really be shorter, and would be less transparemt

**Pull request merged**